File: python/2d-dynamic-programming/edit-distance.py
class Solution:
    def minDistance(self, word1: str, word2: str) -> int:
        cache = {}

        def dfs(i, j):
            if i == len(word1) and j == len(word2):
                return 0
            if i == len(word1) or j == len(word2):
                return abs(i - len(word1)) + abs(j - len(word2))

            if (i, j) in cache:
                return cache[(i, j)]

            if word1[i] == word2[j]:
                cache[(i, j)] = dfs(i + 1, j + 1)
            else:
                cache[(i, j)] = min(dfs(i + 1, j), dfs(i, j + 1), dfs(i + 1, j + 1)) + 1

            return cache[(i, j)]

        return dfs(0, 0)

File: c++/2d-dynamic-programming/edit-distance.cc
class Solution {
public:
    int minDistance(string word1, string word2) {
        std::vector<std::vector<std::optional<int>>> indexes(word1.size(), std::vector<std::optional<int>>(word2.size(), std::nullopt));
        return edit(0, 0, word1, word2, indexes);
    }

private:
    int edit(int i, int j, std::string& word1, std::string& word2, std::vector<std::vector<std::optional<int>>>& indexes) {
        if (i == word1.size() && j == word2.size()) {
            return 0;
        }

        if (i == word1.size() || j == word2.size()) {
            return word1.size() - i + word2.size() - j;
        }

        if (indexes[i][j] != std::nullopt) {
            return indexes[i][j].value();
        }

        if (word1[i] == word2[j]) {
            indexes[i][j] = edit(i + 1, j + 1, word1, word2, indexes);
        } else {
            indexes[i][j] = std::min({edit(i + 1, j, word1, word2, indexes), edit(i + 1, j + 1, word1, word2, indexes), edit(i, j + 1, word1, word2, indexes)}) + 1;
        }

        return indexes[i][j].value();
    }
};
